internal/util/addmergecomment/addmergecomment.go

```diff
@@ -60,15 +102,33 @@ func (amc *AddMergeComment) Filter(object *kyaml.RNode) (*kyaml.RNode, error) {
 		return object, nil
 	}
 	mf := object.Field(kyaml.MetadataField)
-	if mf.IsNilOrEmpty() {
-		// skip adding merge comment if empty metadata
+	if object.GetName() == "" && object.GetNamespace() == "" && len(object.GetLabels()) == 0 {
+		// skip adding merge comment if empty metadata. Since the intermediate annotations always exist,
+		// mf.IsNilOrEmpty cannot tell whether it's empty meta or not.
+		// e.g. Empty meta with internal annotations.
+		//kind: MyKind
+		//spec:
+		//  replicas: 3
+		//metadata:
+		//  annotations:
+		//    config.kubernetes.io/index: '0'
+		//    config.kubernetes.io/path: 'k8s-cli-982798852.yaml'
+		//    internal.config.kubernetes.io/index: '0'
+		//    internal.config.kubernetes.io/path: 'k8s-cli-982798852.yaml'
+		//    internal.config.kubernetes.io/seqindent: 'compact'
+		//    internal.config.kubernetes.io/annotations-migration-resource-id: '0'
 		return object, nil
 	}
-	if strings.Contains(mf.Key.YNode().LineComment, merge.MergeCommentPrefix) {
-		// skip adding merge comment if merge comment is already present
+
+	// Only add merge comment if merge comment does not present
+	if !strings.Contains(mf.Key.YNode().LineComment, merge.MergeCommentPrefix) {
```

Comment:
hmm... I thought the merge-key comment tracks the immediate upstream resource's identifier, but looking this it tracks the root resource's identifier.

Two options:
1) preserve existing behavior (merge-key tracks root resources'id)
2) or change it to use immediate upstream

Both of these will work for update, only difference is the number of descendant nodes affected when root change's name/namespace.

Reply:
I think both options can give identifier functions reliable id and both are more reasonable than the other in certain scenarios. So it defers to future plans on which scenarios are more important.
I'll switch to immediate upstream and list the pros and cons on different options in the PR description.